### ISSUE TITLE
OracleDB: make .hasTable read from ALL_TABLES instead of USER_TABLES

### ISSUE THREAD

**Issue by pablopen** [tgriesser/knex#2662]
# Environment

Knex version: 
Database + version: Oracle 11g
tags: @atiertant. 

# Feature discussion / request

1. If a table is created by another user, and the current user has permissions to view this table, .hasTable returns false because it reads from "USER_TABLES", which only returns tables owned by the user.

2. Changing the query from USER_TABLES to ALL_TABLES will return true, because the table is available for that particular user

My unique concert is that this query from ALL_TABLES will return true, but if you didn't create a synonym, the SELECT will fail, and that's something I don't know how to control in knex. ex:
SELECT * FROM TABLEA;
##ERROR##
SELECT * FROM USERA.TABLEA;
##WORKS##
CREATE SYNONYM TABLEA FOR USERA.TABLEA;
SELECT * FROM TABLEA;
##NOW IT WORKS##

Reference:
User_tables: https://docs.oracle.com/cd/B28359_01/server.111/b28320/statviews_5443.htm#REFRN26286
All_tables: https://docs.oracle.com/cd/B28359_01/server.111/b28320/statviews_2105.htm#REFRN20286

**Comment by pablopen** [comment]
No, USER_TABLES only returns true for tables owned (created) by that user.
 The synonym allows you to query the table using FROM TABLEA (or whatever the synonym is defined) instead of FROM ANOTHERUSER.TABLEA

**Comment by elhigu** [comment]
Ok, thank for the information. I think this change should be fine, I cannot think any drawbacks on having hasTable to query tables from ALL_TABLES.

**Comment by mercmobily** [comment]
I alsi thing that `hasTable()` should check `ALL_TABLES` instead of `USER_TABLES` to support synonyms and tables from other schemas.

The current implementation only queries `USER_TABLES`:

```javascript
// lib/dialects/oracle/schema/oracle-compiler.js:24-32
hasTable(tableName) {
  this.pushQuery({
    sql:
      'select TABLE_NAME from USER_TABLES where TABLE_NAME = ' +
      this.client.parameter(tableName, this.builder, this.bindingsHolder),
    output(resp) {
      return resp.length > 0;
    },
  });
}
```

This means `hasTable()` returns false for:

- Tables accessed through synonyms
- Tables in other schemas that the user has access to

**The problem**

As you and @pablopen explained, Oracle synonyms allow users to access tables from other schemas without prefixing the owner. However, `USER_TABLES` only contains tables owned by the current user, not tables accessible through synonyms or grants.

Changing to `ALL_TABLES` would make `hasTable()` work correctly with synonyms and cross-schema access. As @elhigu confirmed in the discussion, there don't appear to be any drawbacks to this change.

The fix would be straightforward:

```javascript
hasTable(tableName) {
  this.pushQuery({
    sql:
      'select TABLE_NAME from ALL_TABLES where TABLE_NAME = ' +
      this.client.parameter(tableName, this.builder, this.bindingsHolder),
    output(resp) {
      return resp.length > 0;
    },
  });
}
```

This change would:

- Allow `hasTable()` to detect tables accessible through synonyms
- Support cross-schema table access
- Maintain backward compatibility (all tables in USER_TABLES are also in ALL_TABLES)
